File: app/ws_manager.py
"""Gerenciador de conexões WebSocket para atualizações em tempo real
(substitui o STOMP/SimpMessagingTemplate do backend original)."""
import json
from collections import defaultdict
from typing import Dict, Set

from fastapi import WebSocket


class ConnectionManager:
    def __init__(self) -> None:
        self._connections: Dict[str, Set[WebSocket]] = defaultdict(set)

    async def connect(self, event_id: str, websocket: WebSocket) -> None:
        await websocket.accept()
        self._connections[event_id].add(websocket)

    def disconnect(self, event_id: str, websocket: WebSocket) -> None:
        self._connections[event_id].discard(websocket)
        if not self._connections[event_id]:
            self._connections.pop(event_id, None)

    async def broadcast(self, event_id: str, message: dict) -> None:
        dead = []
        for ws in list(self._connections.get(event_id, [])):
            try:
                await ws.send_text(json.dumps(message, default=str))
            except Exception:
                dead.append(ws)
        for ws in dead:
            self.disconnect(event_id, ws)


manager = ConnectionManager()

File: src/utils/wsManager.js
'use strict';
/**
 * Gerenciador de conexões WebSocket para atualizações em tempo real.
 * Equivalente a app/ws_manager.py.
 */
class ConnectionManager {
  constructor() {
    /** @type {Map<string, Set<import('ws')>>} */
    this._connections = new Map();
  }

  connect(eventId, ws) {
    if (!this._connections.has(eventId)) {
      this._connections.set(eventId, new Set());
    }
    this._connections.get(eventId).add(ws);
  }

  disconnect(eventId, ws) {
    const set = this._connections.get(eventId);
    if (!set) return;
    set.delete(ws);
    if (set.size === 0) {
      this._connections.delete(eventId);
    }
  }

  broadcast(eventId, message) {
    const set = this._connections.get(eventId);
    if (!set) return;
    const payload = JSON.stringify(message);
    const dead = [];
    for (const ws of set) {
      try {
        if (ws.readyState === ws.OPEN) {
          ws.send(payload);
        } else {
          dead.push(ws);
        }
      } catch (err) {
        dead.push(ws);
      }
    }
    for (const ws of dead) {
      this.disconnect(eventId, ws);
    }
  }
}

const manager = new ConnectionManager();

module.exports = { manager };
